### ISSUE TITLE
experiment: fine-tuned LLM transcript repair layer (CW decoder)

### ISSUE THREAD

**Issue by rtreit** [rtreit/logripper#422]
## Summary

Explore fine-tuning a small LLM (e.g. GPT-5-nano, Phi-3, Qwen-2.5-3B, Llama-3.2-3B) as a **transcript repair layer** that runs after the classical/neural CW decoder. Input is the noisy decoded string + context (WPM, SNR, optional priors); output is a corrected transcript with confidence.

This is a post-processing layer, not a replacement for the acoustic decoder. It only sees text, never audio.

## Motivation

The bake-off results to date (PRs #410-#421) show that:
- Per-element classification has a hard ceiling driven by acoustic ambiguity (E/T/I/M/S confusion under SNR collapse)
- Class-naive language models regress (PR #421 bigram-corpus): below lambda~=1 the LM cannot flip anything; above, it damages legitimate callsigns
- A repair layer that *understands ham conventions* (callsigns, contest exchanges, POTA/SOTA park IDs, RST, grid squares, CQ ... DE ... K shape) is exactly what's missing

The charlm experiment (PR #410 writeup) hinted at this: a narrow callsign-join pass actually helped WA6MOW. A fine-tuned LM generalizes that pattern to every ham idiom for free.

## Why this fits the architecture

- Lives behind a gRPC `TranscriptRepair` service with unique `RepairTranscriptRequest/Response` envelopes (per the proto 1-1-1 rule)
- Engine remains decoder-agnostic; repair is a separate stage easy to A/B-test, disable, or swap implementations (cloud API vs local model)
- No change to the live keyboard-first UX path: classical decoder emits text in real time; repair runs async and updates the row ~200-500 ms later

## Why it should work

- Training pairs are essentially free:
  - ARRL corpus (1,576 chunks, ground-truth text) -> run our viterbi decoder over it -> ~17.84 h of (noisy, clean) pairs
  - Augmented corpus from #augment-arrl (forthcoming) -> orders of magnitude more pairs at known SNR/Watterson conditions
  - LoTW / QRZ / POTA logs -> millions of real callsign + exchange examples for vocabulary grounding
  - Synthetic confusables: take clean truth, apply known CW-confusable substitutions (E<->I<->S, T<->M<->O, dit-merge, dah-split) to manufacture infinite pairs
- Sample size: ~5-20k pairs is enough for a small fine-tune. Generatable in an afternoon once augment-arrl lands.

## Proposed input shape

```
decoded: "CQPOTADEWA6M0W CQPOTA DE WA6M0W K"
context:
  wpm: 22
  snr_db: 8
  pitch_hz: 704
  callsign_priors: ["WA6MOW", "K7ABC", ...]   # optional, from local logbook
output:
  repaired: "CQ POTA DE WA6MOW CQ POTA DE WA6MOW K"
  confidence: 0.91
  edits: [{span: "WA6M0W", replaced: "WA6MOW", reason: "callsign-correction"}, ...]
```

## Risks (pre-identified from bake-off failures)

- **Hallucinated callsigns** - charlm regressed by correcting rare-but-real callsigns toward common ones. Mitigations: constrained decoding against an actual callsign database (LoTW dump?), high abstention threshold, return edits with reasons so user can reject.
- **Latency** - not viable for live keyboard-first contest UX. Two-tier: classical for live text, LM repair runs async behind it.
- **Noise-only hallucination** - model must not emit text from `???` input. Train with explicit abstain examples (noise-only -> empty).
- **Vocabulary drift** - model trained on US-heavy data may regress on JA/EU callsigns, weird POTA park codes, etc. Need geographic balance in training set.

## Acceptance criteria for a first experiment

1. Train a small model (3-7B local or GPT-5-nano API) on >=5k pairs from ARRL + augmented corpus
2. Evaluate on the 6-sample bench used by the rest of the bake-off, plus the adversarial-suite (PR #417) weak-prefix and ghost-cascade suites
3. Required gates:
   - Mean CER on real bench: improves over viterbi (current best 0.202)
   - WA6MOW WER: <= 0.20 (current 0.60)
   - cq-pota-aa6pw CER: <= 0.15 (current 0.167)
   - Adversarial noise-only suite: must remain 100% empty (no hallucination)
   - Adversarial weak-prefix suite: improves over viterbi without callsign damage
4. Honest report on:
   - Latency p50/p95
   - Disk + memory footprint of local inference path
   - Cost per 1k repairs for any cloud API path

## Sequence

This depends on:
- #augment-arrl (PR TBD) - produces the augmentation pipeline that lets us generate training data at scale
- #wpm-seed-fix (PR TBD) - gets the classical decoder to a stable acoustic baseline before we layer LM repair on top

Once both have landed, this becomes the next experiment in the queue.

## Open design questions (resolve when scheduling)

1. Cloud API (GPT-5-nano) vs local fine-tune (Phi-3 / Qwen / Llama)? Probably both; A/B in DebugHost.
2. Constrained decoding against a callsign database, or unconstrained with confidence gating?
3. Edit-level output (`span -> replacement`) or full-string rewrite? Edit-level is more debuggable and lets the user accept/reject per change.
4. Real-time streaming repair (token-by-token) vs end-of-region only? Latency vs UX trade-off.
